### ISSUE TITLE
Type checker mistakes predicate instance for function application and crashes

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#127]
[Issue created by mschwerhoff: 2015-12-03]
[Last updated on bitbucket: 2019-08-28]

[Comment created by mschwerhoff: 2015-12-03]
See `wands\regression\potential_unsoundness.sil`

[Comment created by mschwerhoff: 2017-12-06]
The degree to which wands are treated syntactically (upon lookup), i.e. which subexpressions are treated as holes/are compared semantically, should also be clarified. It should also be documented, e.g. in the tutorial. See also issue #230.

[Comment created by alexander_summers: 2019-08-28]
See the tutorial issue: [https://bitbucket.org/viperproject/tutorial/issues/3/clarify-how-magic-wand-instances-are](https://bitbucket.org/viperproject/tutorial/issues/3/clarify-how-magic-wand-instances-are)

[Comment created by alexander_summers: 2019-08-28]
We should make sure the interesting cases are handled in the test suite, for the sake of closing this particular \(Silver\) issue

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2016-04-08 09:41

Probably the same problem:

```
#!text

method f(n: Int) returns (res: Int)
{
	if (n > 100) { res := n - 10}
	else {
		res := f(f(n + 11))
	}
}

Exception in thread "main" java.lang.ClassCastException: viper.silver.parser.PMethod cannot be cast to viper.silver.parser.PAnyFunction
	at viper.silver.parser.TypeChecker.check(Resolver.scala:513)
	at viper.silver.parser.TypeChecker.checkTopTyped(Resolver.scala:432)
	at viper.silver.parser.TypeChecker.check(Resolver.scala:429)
	at viper.silver.parser.TypeChecker$$anonfun$check$19.apply(Resolver.scala:211)
	at viper.silver.parser.TypeChecker$$anonfun$check$19.apply(Resolver.scala:210)
        ...
```

The following also results in a `ClassCastException`:

    predicate P(n: Int)
    ...
    res := f(P(n + 11))

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2016-04-12 12:18

See also #49, #137

**Comment by fpoli** [comment]
> **@meilers** commented on 2016-07-29 12:47

Resolved in pull request #54.
